packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -5779,137 +5781,167 @@ namespace Tpetra {
-  computeCrsPaddingNew (const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
-                        const size_t numSameIDs,
-                        const Kokkos::DualView<const local_ordinal_type*,
-                          buffer_device_type>& permuteToLIDs,
-                        const Kokkos::DualView<const local_ordinal_type*,
-                          buffer_device_type>& permuteFromLIDs)
+  computeCrsPaddingForSameIDs (Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>& padding,
+                               const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
+                               const size_t numSameIDs,
+                               const bool padAll) const
   {
     using LO = LocalOrdinal;
     using GO = GlobalOrdinal;
     using execution_space = typename device_type::execution_space;
-    const char tfecfFuncName[] = "computeCrsPaddingNew: ";
+    const char tfecfFuncName[] = "computeCrsPaddingForSameIds: ";
 
     execution_space::fence ();
 
-    // Resize row pointers and indices to accommodate incoming data
+    using insert_result =
+      typename Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>::insert_result;
+
+    // Compute extra capacity needed to accommodate incoming data
+    Teuchos::Array<GO> src_row_inds;
+    Teuchos::Array<GO> tgt_row_inds;
     const map_type& src_row_map = * (source.getRowMap ());
-    using padding_type = Kokkos::UnorderedMap<LO, size_t, device_type>;
-    padding_type padding (numSameIDs + permuteFromLIDs.extent (0));
-    for (LO tgtid = 0; tgtid < static_cast<LO> (numSameIDs); ++tgtid) {
-      const GO srcgid = src_row_map.getGlobalElement (tgtid);
-      auto how_much_padding = source.getNumEntriesInGlobalRow (srcgid);
-      auto result = padding.insert (tgtid, how_much_padding);
-      // FIXME (mfh 09 Apr 2019) Kokkos::UnorderedMap is allowed to
-      // fail even if the user did nothing wrong.  We should actually
-      // have a retry option.  I just copied this code over from
-      // computeCrsPadding.
+    for (LO tgt_lid = 0; tgt_lid < static_cast<LO> (numSameIDs); ++tgt_lid) {
+      const GO src_gid = src_row_map.getGlobalElement(tgt_lid);
+      auto num_src_entries = source.getNumEntriesInGlobalRow(src_gid);
+
+      if (num_src_entries == 0)
+        continue;
+
+      insert_result result;
+      const GO tgt_gid = rowMap_->getGlobalElement(tgt_lid);
+      if (padAll) {
+        result = padding.insert(tgt_lid, num_src_entries);
+      }
+      else {
+        size_t check_row_length = 0;
+        src_row_inds.resize(num_src_entries);
+        source.getGlobalRowCopy(src_gid, src_row_inds(), check_row_length);
+
+        auto num_tgt_entries = this->getNumEntriesInGlobalRow(tgt_gid);
+        tgt_row_inds.resize(num_tgt_entries);
+        this->getGlobalRowCopy(tgt_gid, tgt_row_inds(), check_row_length);
+
+        size_t how_much_padding = 0;
+        for (const auto& src_row_ind : src_row_inds) {
```

Comment:
References are bigger than values in this case, so you might as well just use `auto` instead of `const auto&` here.